ietfdata/tools/affiliations.py

```diff
@@ -0,0 +1,152 @@
+# TODO: Add licensing
+
+import sys
+import csv
+import json 
+import copy
+
+import datetime
+from datetime import timedelta
+from typing import Optional
+
+from ietfdata.datatracker     import *
+from ietfdata.datatracker_ext import *
+from ietfdata.mailarchive2    import *
+
+# load affiliation_mappings
+# import affiliations_map #hugo
+import ietfdata.tools.affiliation_mapping_dictionary as afmap #yangjun
+# affiliations.py --- script to generate extract affiliations
+# Mappings generated:
+# 1. raw affiliation -> normalised affiliation mappings
+# 2. email_domain -> normalised affiliations mappings
+# 3. identity -> start year-month, end year-month, affiliation mapping
+
+# Affiliation Entry Class
+class AffiliationEntry:
+    start_date  : datetime.date
+    end_date    : Optional[datetime.date]
+    names : list[str]
+    
+    def __init__(self, affiliation_str:str, start_date:datetime.date, end_date:Optional[datetime.date],names_list:Optional[list[str]]):
+        if names_list is not None:
+            self.names=copy.deepcopy(names_list)
+        else:
+            self.names = cleanup_affiliation(affiliation_str)
+        self.start_date = start_date
+        self.end_date = end_date
+    
+    def set_end_date(self, new_end_date:datetime.date):
+        self.end_date = new_end_date
+    
+    def set_affiliation_names(self, affiliation:str):
+        self.namse=cleanup_affiliation(affiliation)
+    
+    def match_names(self, names:list[str]):
+        for name in names:
+            if name not in self.names:
+                return False
+        return True
+    
+    def __str__(self):
+        return_str = '{"names":['
+        for name in self.names:
+            return_str+=f'"{name}",'
+        return_str =  return_str.rstrip(',')
+        return_str += "],"
+        return_str += f'"start_date":"{self.start_date}","end_date":"{self.end_date}"}}'
+        return return_str
+
+# Affiliation mapping class
+class AffiliationMap:
+    identifiers : list[str]
+    affiliations: list[AffiliationEntry]
+    
+    def __init__(self,identifiers:list[str],affiliations:list[AffiliationEntry]):
+        self.identifiers = copy.deepcopy(identifiers)
+        self.affiliations = copy.deepcopy(affiliations)
+    
+    def add_identifier(self, identifier:str):
+        self.identifiers.append(identifier)
+    
+    def add_affiliation(self, aff_entry:AffiliationEntry):
+        # TODO: Go through the timeline, insert the entry 
+        new_date = aff_entry.start_date
+        for affil in self.affiliations:
+            i = self.affiliations.index(affil)
+            if new_date <= affil.start_date:
+                self.affiliations.insert(i,aff_entry)
+                return #inserted entry
+        self.affiliations.append(aff_entry)
+        return
+            
+    def consolidate(self):
+        # TODO: Go through the timeline, consolidate the history
+        # This should only be run if and only if everything has been scraped
```

Comment:
Does anything break if this is run more than once, or is it just inefficient to do so?